### DISCUSSION TITLE
Including LAV Filters pyinstaller

### DISCUSSION THREAD

**Discussion by DL854359743057** [pyinstaller/pyinstaller#6610] [Help]
I am trying to deploy a pyqt5 application with pyinstaller on Windows.
The main problem now is that some functions are dependent on LAV Filters.
So after building the executable and running the program on the building machine it works perfect. However, on a different machine I get following error:
`DirectShowPlayerService::doRender: Unresolved Error Code 0x80040226 (IDispatch error #102)`
This is because LAV Filters are not installed on this machine.
I tried to include the LAV Filters directory and path in the .spec file as following:
`pathex=['C:\\Program Files (x86)\\LAV Filters']`
and
`datas=[('C:\\Program Files (x86)\\LAV Filters', 'LAV Filters')]`
However, this does not solve the problem.

An ugly partial solution would be to include the LAV Filters installer in the onefile and then run the installer as the first part of the python script. 
```
import os
os.system('LAVFilters-0.76.exe')
```
But some users do not have admin rights on their machines, so this approach fails for my use case.

**So my question is, how do I properly include LAV Filters in pyinstaller onefile?**

**Comment by rokm** [comment]
Just collecting `LAV Filters` directory probably does nothing if those functions are searching for them in specific location(s). 

If that location is fixed and static (e.g., hardcoded path to `C:\\Program Files (x86)\\LAV Filters`, a path stored in registry, etc.), then you probably won't be able to bundle them with your application, and the user will have to install them separately.

If you can manipulate the search path (via an environment variable, or a string passed to function calls, etc.), then you can adjust it accordingly at the start of your program (in a frozen-application-specific codepath), and have it point at `os.path.join(sys._MEIPASS), 'LAV Filters')`.

**Reply by DL854359743057** [reply]
Thanks for your reply!

I still can not figure it out... 
So I installed LAV Filters on an additional machine. I could confirm that the problem resolved, when installing LAV Filters.
I also checked the Windows environment variables after installation. It seems that LAV Filters did not add any new entries or adjusted the PATH variable. So the environment variables did not change with the installation.

Then I uninstalled LAV Filters and of cause the usual error occurred when starting my pyinstaller generated .exe.
Before I uninstalled I copied the folder `C:\\Program Files (x86)\\LAV Filters` to the Desktop.
Now I copied the `LAV Filters` folder from the Desktop back to the installation path `C:\\Program Files (x86)\\LAV Filters` .
Then I tried to run the .exe again and the usual error occurred :(
This tells me, that there are some additional things going on during the installation, so copying folders does not work...
I also tried to add this so my code, but this did not yield success:
```
def resource_path(relative_path):
    try:
        base_path = sys._MEIPASS
    except Exception:
        base_path = os.path.abspath(".")
    return os.path.join(base_path, relative_path)

os.environ['PATH'] = resource_path('LAV Filters')
os.environ['PATH'] += resource_path('LAV Filters/x64')
os.environ['PATH'] += resource_path('LAV Filters/x86')
```

The LAV filters readme (https://github.com/Nevcairiel/LAVFilters/blob/master/README.txt)
talks about `Register (install_*.bat files)` during installation.

So I really do not know where to go from here, because manual installation is not an option for my project. 
Any ideas?

**Reply by rokm** [reply]
If you check those `install_*.bat` files, you'll see that the filters are registered via `regsvr32.exe`:

```
cd /d "%~dp0"
regsvr32.exe LAVVideo.ax
```

**Reply by rokm** [reply]
So I guess this won't work for your non-admin users anyway, because running `regsvr32` also requires admin privileges.

**Reply by DL854359743057** [reply]
Ok, I understand.
Thank you for your effort and time!
